### PR TITLE
[Perf] Optimize compute maxsim using batched version, 3.2% E2E throughput improvement

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

A follow up PR for https://github.com/vllm-project/vllm/pull/36159, now we are using a batched version to optimize.

CC: @noooop @DarkLight1337

### Acc

Covered in unit tests

```bash
tests/entrypoints/pooling/score/test_online_colbert.py::TestColBERTOnline::test_score
tests/entrypoints/pooling/score/test_online_colbert.py::TestColBERTOnline::test_rerank
tests/entrypoints/pooling/score/test_online_colbert.py::TestColBERTOnline::test_rerank_top_n
...
```

### Perf

`vllm serve   --model jinaai/jina-colbert-v2   --runner pooling   --port 9256   --enforce-eager   --max-model-len 4096   --max-num-batched-tokens 4096   --disable-log-stats   --hf-overrides '{"architectures": ["ColBERTJinaRobertaModel"]}'   --trust-remote-code`

`vllm bench serve   --model jinaai/jina-colbert-v2   --backend vllm-rerank   --endpoint /v1/rerank   --host 127.0.0.1 --port 9256   --dataset-name random-rerank   --num-prompts 2000   --request-rate inf   --max-concurrency 64   --seed 0   --random-input-len 2048   --random-range-ratio 0.5   --percentile-metrics e2el   --metric-percentiles 50,95,99`

```bash
# now
============ Serving Benchmark Result ============
Successful requests:                     2000      
Failed requests:                         0         
Maximum request concurrency:             64        
Benchmark duration (s):                  11.69     
Total input tokens:                      4116113   
Request throughput (req/s):              171.08    
Total token throughput (tok/s):          352098.39 
----------------End-to-end Latency----------------
Mean E2EL (ms):                          368.25    
Median E2EL (ms):                        367.35    
P50 E2EL (ms):                           367.35    
P95 E2EL (ms):                           410.05    
P99 E2EL (ms):                           544.24    
==================================================
# main
============ Serving Benchmark Result ============
Successful requests:                     2000      
Failed requests:                         0         
Maximum request concurrency:             64        
Benchmark duration (s):                  12.08     
Total input tokens:                      4116113   
Request throughput (req/s):              165.63    
Total token throughput (tok/s):          340878.45 
----------------End-to-end Latency----------------
Mean E2EL (ms):                          380.50    
Median E2EL (ms):                        380.24    
P50 E2EL (ms):                           380.24    
P95 E2EL (ms):                           407.26    
P99 E2EL (ms):                           452.19    
==================================================
```

### CODE REVIEW

**Review by DarkLight1337 — COMMENTED** [review]
Is the benchmark for "main" branch with or without #36159?

File: vllm/v1/worker/gpu/pool/late_interaction_runner.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 from vllm.v1.pool.late_interaction import (
     LATE_INTERACTION_MODE_CACHE_QUERY,
     LATE_INTERACTION_MODE_SCORE_DOC,
-    compute_maxsim_score,
+    compute_maxsim_scores,
```
Comment (DarkLight1337) [inline vllm/v1/worker/gpu/pool/late_interaction_runner.py:12]: Can we remove the old `compute_maxsim_score`? Or is it for reference

File: vllm/v1/pool/late_interaction.py [hunk 2]
```diff
@@ -62,3 +63,81 @@ def compute_maxsim_score(
...
+            (batch_size, max_q, dim), dtype=torch.float32, device=device
+        )
+        d_batch = torch.zeros(
+            (batch_size, max_d, dim), dtype=torch.float32, device=device
+        )
+        q_mask = torch.zeros((batch_size, max_q), dtype=torch.bool, device=device)
+        d_mask = torch.zeros((batch_size, max_d), dtype=torch.bool, device=device)
+
+        # copy to padded tensors
+        for i, (q_emb, d_emb) in enumerate(zip(batch_q, batch_d)):
```
[the remark this answers is not in the document]
Comment (yewentao256) [reply vllm/v1/pool/late_interaction.py:127]: Error report

File: vllm/v1/worker/gpu/pool/late_interaction_runner.py (hunk 1 above)
Comment (yewentao256) [reply vllm/v1/worker/gpu/pool/late_interaction_runner.py:12]: It is used for reference in unit test

**Review by DarkLight1337 — APPROVED** [review]
LGTM, but would like @noooop to confirm
